In [1]:
import pandas as pd

On the instructions' recommendation, I tried to "simply use pandas to read the table into a pandas dataframe" ...

In [2]:
toronto_hoods=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

... I didn't get a DataFrame ...

In [3]:
toronto_hoods

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

After much trial and error, exploring what datatype this data was--I found it was a List-- I tried to just seet what the first element of this list was ...

In [4]:
toronto_hoods[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                         Not assigned  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
10                                        Not assigned  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                        Not assigned  
16                                        Not assigned  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                        Not assigned  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                        Not assigned  
25                                        Not assigned  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                        Not assigned  

... that looks like a DataFrame!

Checking ...

In [5]:
type(toronto_hoods[0])

pandas.core.frame.DataFrame

So it is a DataFrame ... ok.

On with the project ...

In [6]:
# Shortening things up ...
tHoods=toronto_hoods[0]
tHoods.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

The instructions state I should "Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned."

So ...

In [7]:
tHoods=tHoods[tHoods['Borough']!='Not assigned']
tHoods.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

I'm assuming no missing or useless values remain, given the heads-up from the instructions.

Now, I'm trying to use the Shape method ...

In [8]:
tHoods.shape()

TypeError: 'tuple' object is not callable

...--and that result is despite the fact that tHoods ***is*** a DataFrame ...

In [9]:
type(tHoods)

pandas.core.frame.DataFrame

... it's not yielding a result when the Shape method is used.

We'll see how much that affect the rest of the project.

tHoods' dimensions are knowable, still:

In [10]:
tHoods.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 2 to 178
Data columns (total 3 columns):
Postal Code     103 non-null object
Borough         103 non-null object
Neighborhood    103 non-null object
dtypes: object(3)
memory usage: 3.2+ KB


The "DataFrame" being analyzed is 180 x 3, apparently.

Now we look at geospatial data about tHoods. We'll call it tGeo.

In [15]:
tGeo=pd.read_csv('https://cocl.us/Geospatial_data')
tGeo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [ ]:
Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

Just make sure:

to add enough Markdown cells to explain what you decided to do and to report any observations you make.
to generate maps to visualize your neighborhoods and how they cluster together.

Taking a look around the neighborhood (mathematically) ...

In [16]:
tGeo.describe()

Latitude   Longitude
count  103.000000  103.000000
mean    43.704608  -79.397153
std      0.052463    0.097146
min     43.602414  -79.615819
25%     43.660567  -79.464763
50%     43.696948  -79.388790
75%     43.745320  -79.340923
max     43.836125  -79.160497

In a manner of speaking, the "heart" of Toronto is ... 43.7 lat, -79.4 long ...

Space Needle?

In [17]:
tGeo.isnull()

Postal Code  Latitude  Longitude
0          False     False      False
1          False     False      False
2          False     False      False
3          False     False      False
4          False     False      False
5          False     False      False
6          False     False      False
7          False     False      False
8          False     False      False
9          False     False      False
10         False     False      False
11         False     False      False
12         False     False      False
13         False     False      False
14         False     False      False
15         False     False      False
16         False     False      False
17         False     False      False
18         False     False      False
19         False     False      False
20         False     False      False
21         False     False      False
22         False     False      False
23         False     False      False
24         False     False      False
25         False     False      False
26         False     False      False
27         False     False      False
28         False     False      False
29         False     False      False
..           ...       ...        ...
73         False     False      False
74         False     False      False
75         False     False      False
76         False     False      False
77         False     False      False
78         False     False      False
79         False     False      False
80         False     False      False
81         False     False      False
82         False     False      False
83         False     False      False
84         False     False      False
85         False     False      False
86         False     False      False
87         False     False      False
88         False     False      False
89         False     False      False
90         False     False      False
91         False     False      False
92         False     False      False
93         False     False      False
94         False     False      False
95         False     False      False
96         False     False      False
97         False     False      False
98         False     False      False
99         False     False      False
100        False     False      False
101        False     False      False
102        False     False      False

[103 rows x 3 columns]

Looks good (to me ..)

Look Ma--I'm making maps!

Folium ... Folium is makin maps ...

In [18]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    ------------------------------------------------------------
                       

A base map of tHoods, which I'll mark up with ... markers ... later, to show how close tHoods are (their clustering)--and I'm using that heart spot I found earlier for a location center:

In [20]:
tHoodsMap=folium.Map(location=[43.704608, -79.397153], zoom_start=12)
tHoodsMap

In [37]:
lats=tGeo.iloc[:,1]
lats[:5]

0    43.806686
1    43.784535
2    43.763573
3    43.770992
4    43.773136
Name: Latitude, dtype: float64

In [36]:
lons=tGeo.iloc[:,2]
lons[:5]

0   -79.194353
1   -79.160497
2   -79.188711
3   -79.216917
4   -79.239476
Name: Longitude, dtype: float64

In [40]:
tHoodMarkers=folium.map.FeatureGroup()

for lat, lon in zip(lats, lons):
    tHoodMarkers.add_child(
        folium.features.CircleMarker(
            [lat, lon],
            radius=5, # define how big you want the circle markers to be
            color='yellow',
            fill=True,
            fill_color='blue',
            fill_opacity=0.6
                                    )
                           )

tHoodsMap.add_child(tHoodMarkers)

Somewhere ... by the sea, somewhere, where I won't be ...

Toronto's on the water?

Way cold, thinkin bout it.

I hear it's lovely--and ***clean*** ...